google/cloud/spanner/read_partition.h

```diff
@@ -0,0 +1,158 @@
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <google/spanner/v1/spanner.pb.h>
+#include <string>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+class ReadPartition;
+
+/**
+ * Serializes an instance of `ReadPartition` for transmission to another
+ * process.
+ *
+ * @param read_partition - instance to be serialized.
+ *
+ * @par Example
+ *
+ * @code
+ * std::vector<spanner::ReadPartition> partitions =
+ *   spanner_client.PartitionRead(...);
+ * for (auto const& partition : partitions) {
+ *   auto serialized_partition = spanner::SerializeReadPartition(partition);
+ *   if (serialized_partition.ok()) {
+ *     SendToRemoteMachine(*serialized_partition);
+ *   }
+ * }
+ * @endcode
+ */
+StatusOr<std::string> SerializeReadPartition(
+    ReadPartition const& read_partition);
+
+/**
+ * Deserializes the provided string into a `ReadPartition`, if able.
+ *
+ * Returned `Status` should be checked to determine if deserialization was
+ * successful.
+ *
+ * @param serialized_read_partition - string representation to be deserialized.
+ *
+ * @par Example
+ *
+ * @code
+ * std::string serialized_partition = ReceiveFromRemoteMachine();
+ * spanner::ReadPartition partition =
+ *   spanner::DeserializeReadPartition(serialized_partition);
+ * auto rows = spanner_client.Read(partition);
+ * @endcode
+ */
+StatusOr<ReadPartition> DeserializeReadPartition(
+    std::string const& serialized_read_partition);
+
+// Internal implementation details that callers should not use.
+namespace internal {
+
+ReadPartition MakeReadPartition(std::string transaction_id,
+                                std::string session_id,
+                                std::string partition_token,
+                                std::string table_name, KeySet key_set,
+                                std::vector<std::string> column_names,
+                                ReadOptions read_options = {});
+}  // namespace internal
+
+/**
+ * The `ReadPartition` class is a regular type that represents a single
+ * slice of a parallel Read operation.
+ *
+ * Instances of `ReadPartition` are created by `Client::PartitionRead`. Once
+ * created, `ReadPartition` objects can be serialized, transmitted to separate
+ * process, and used to read data in parallel using `Client::Read`.
+ */
+class ReadPartition {
+ public:
+  /**
+   * Constructs an instance of `ReadPartition` that does not specify any table
+   * or columns to be read.
+   */
+  ReadPartition() = default;
+
+  /// @name Copy and move.
+  ///@{
+  ReadPartition(ReadPartition const&) = default;
+  ReadPartition(ReadPartition&&) = default;
+  ReadPartition& operator=(ReadPartition const&) = default;
+  ReadPartition& operator=(ReadPartition&&) = default;
+  ///@}
+
+  std::string table_name() const { return proto_.table(); }
+  std::vector<std::string> column_names() const {
+    auto columns = proto_.columns();
```

Comment:
Will this make a copy of all the columns here? If so, you may want to avoid that extra copy w/ `auto const&`.